Sources/AuthFoundation/Migration/Migration.swift

```diff
@@ -0,0 +1,82 @@
+//
+// Copyright (c) 2022-Present, Okta, Inc. and/or its affiliates. All rights reserved.
+// The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+//
+// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//
+// See the License for the specific language governing permissions and limitations under the License.
+//
+
+import Foundation
+
+public enum SDKVersionMigrationError: Error {
+    case cannotCreateMigrator
+    case configurationNotFound
+    case configurationInvalid
+}
+
+extension SDKVersion {
+    /// Determines whether or not some user data needs to be migrated.
+    ///
+    /// This may be if a user has upgraded to a newer version of the SDK.
+    public static var isMigrationNeeded: Bool {
+        Migration.shared.needsMigration
+    }
+    
+    /// Migrates user data, if necessary.
+    public static func migrateIfNeeded() throws {
+        guard isMigrationNeeded else {
+            return
+        }
+        
+        try Migration.shared.migrate()
+    }
+    
+    /// Registers an SDK version migrator for use within a migration process.
+    ///
+    /// Version migrators are utilized to migrate user data on an as-needed basis.
+    /// - Parameter migrator: Migrator to register.
+    public static func register(migrator: SDKVersionMigrator) {
+        Migration.registeredMigrators.append(migrator)
+    }
+    
+    /// Namespace used for a variety of version migration agents.
+    public class Migration {
+        fileprivate static var hasMigrated: Bool = false
+        static var shared: Migration = {
```

Comment:
I want to deallocate this shared instance after migration has completed, but it introduces some interesting race conditions.  I might change this later, but for now I don't want to prematurely optimize things.